File: src/main/java/red/sukun1899/service/IndexService.java
package red.sukun1899.service;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import red.sukun1899.AppConfig;
import red.sukun1899.model.Index;
import red.sukun1899.repository.IndexRepository;

import java.util.List;
import java.util.stream.Collectors;

/**
 * @author su-kun1899
 */
@Service
public class IndexService {
    private AppConfig appConfig;
    private IndexRepository indexRepository;

    public IndexService(AppConfig appConfig, IndexRepository indexRepository) {
        this.appConfig = appConfig;
        this.indexRepository = indexRepository;
    }

    @Transactional(readOnly = true)
    public List<Index> get(String tableName) {
        return indexRepository.selectByTableName(appConfig.getSchemaName(), tableName).stream()
                .sorted((index1, index2) -> {
                    if (index1.getCategory().order() - index2.getCategory().order() != 0) {
                        return index1.getCategory().order() - index2.getCategory().order();
                    }
                    return index1.getName().compareTo(index2.getName());
                })
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/red/sukun1899/shishamo/service/IndexService.kt
package red.sukun1899.shishamo.service;

import org.springframework.boot.autoconfigure.jdbc.DataSourceProperties
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import red.sukun1899.shishamo.model.Index
import red.sukun1899.shishamo.repository.IndexRepository

/**
 * @author su-kun1899
 */
@Service
class IndexService(
        val dataSourceProperties: DataSourceProperties,
        val indexRepository: IndexRepository
) {
    @Transactional(readOnly = true)
    fun get(tableName: String): List<Index> = indexRepository.selectByTableName(dataSourceProperties.schema, tableName)
            .sortedWith(Comparator { (name1, _, category1), (name2, _, category2) ->
                if (category1.order != category2.order) category1.order - category2.order
                else name1.compareTo(name2)
            })
}
